zio-http/src/main/scala/zio/http/Flash.scala

```diff
@@ -0,0 +1,390 @@
+     * Gets an `A` from the backend-based flash-scope or fails with a
+     * `Throwable`.
+     */
+    def flash[A](request: Request, flash: Flash[A]): IO[Throwable, A]
+
+    /**
+     * Gets an `A` from the backend-based flash-scope and provides a fallback.
+     */
+    def flashOrElse[A](request: Request, flash: Flash[A])(orElse: => A): UIO[A] =
+      self.flash(request, flash) <> ZIO.succeed(orElse)
+
+    /**
+     * Adds flash values to the backend-based flash-scope and returns a workflow
+     * with an updated `Response`.
+     */
+    def addFlash[A](response: Response, setter: Flash.Setter[A]): UIO[Response]
+
+    /**
+     * Optionally adds flash values to the backend-based flash-scope and returns
+     * a workflow with an updated `Response`.
+     */
+    def addFlash[A](response: Response, setterOpt: Option[Flash.Setter[A]]): UIO[Response] =
+      setterOpt.fold(ZIO.succeed(response))(self.addFlash(response, _))
+  }
+
+  object Backend {
+
+    private case class Impl(ref: Ref[Map[UUID, Map[String, String]]]) extends Backend {
+      override def flash[A](request: Request, flash: Flash[A]): IO[Throwable, A] =
+        for {
+          flashId <- ZIO.from(Flash.run(Flash.getUUID(flashIdName), request))
+          a       <- ref.modify { map =>
+            Flash.run(flash, map.get(flashId).getOrElse(Map.empty)) match {
+              case value @ Right(_) => value -> (map - flashId)
+              case value @ Left(_)  => value -> map
+            }
+          }.flatMap(ZIO.from(_))
+        } yield a
+
+      override def addFlash[A](response: Response, setter: Setter[A]): UIO[Response] = {
+        val map = Flash.Setter.run(setter, Map.empty)
+        for {
+          flashId       <- zio.Random.nextUUID
+          setterFlashId <- ref.update(in => in + (flashId -> map)).as(Flash.setValue(flashIdName, flashId))
+        } yield response.addFlash(setterFlashId)
+      }
+    }
+
+    val layer: ULayer[Backend] = ZLayer(Ref.make(Map.empty[UUID, Map[String, String]]).map(Impl.apply))
+
+    private val flashIdName = "flashId"
+
+  }
+
+  sealed trait Setter[A] { self =>
+
+    /**
+     * Combines setting this flash value with another setter `that`.
+     */
+    def ++[B](that: => Setter[B]): Setter[(A, B)] = Setter.Concat(self, that)
+  }
+
+  private[http] object Setter {
+
+    case object Empty extends Flash.Setter[Unit]
+
+    case class SetValue[A](schema: Schema[A], key: String, a: A) extends Flash.Setter[A]
+
+    case class Concat[A, B](left: Setter[A], right: Setter[B]) extends Flash.Setter[(A, B)]
+
+    def run[A](setter: Setter[A]): Cookie.Response =
+      Cookie.Response(
+        Flash.COOKIE_NAME,
+        URLEncoder.encode(
+          JsonCodec.jsonEncoder(Schema[Map[String, String]]).encodeJson(run(setter, Map.empty)).toString,
+          java.nio.charset.Charset.defaultCharset.toString.toLowerCase,
+        ),
+      )
+
+    def run[A](setter: Setter[A], map: Map[String, String]): Map[String, String] = {
```

Comment:
We could inline this `run` function into the previous one, no?

```scala
def run[A](setter: Setter[A]): Cookie.Response = {
    def loop(setter: Setter[A], map: mutable.Map[String, String]): Map[String, String] = 
      setter match {
          case SetValue(schema, key, a) =>
            map += (key -> JsonCodec.jsonEncoder(schema).encodeJson(a).toString)
          case Concat(left, right)      =>
            loop(right, loop(left, map))
          case Empty                    => map.toMap
        }

      Cookie.Response(
        Flash.COOKIE_NAME,
        URLEncoder.encode(
          JsonCodec.jsonEncoder(Schema[Map[String, String]]).encodeJson(loop(setter, mutable.Map.empty)).toString,
          java.nio.charset.Charset.defaultCharset.toString.toLowerCase,
      )
    )         
}
```

Reply:
Internally there both `run` operations are used